model/riscv_vmem.sail

```diff
@@ -0,0 +1,855 @@
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// PRIVATE
+val      undefined_SV_Params : unit -> SV_Params
+
+function undefined_SV_Params() = sv32_params
+
+// ****************************************************************
+// Fields of virtual addresses
+
+// PRIVATE: Extract full VPN field from VA
+function vpns_of_va(sv_params : SV_Params,
+                    va        : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.va_size_bits) - zero_extend(0b1);
+    (va & mask) >> PAGESIZE_BITS
+}
+
+// PRIVATE: Extract VPN[level] from  VA
+function vpn_j_of_va(sv_params : SV_Params,
+                     va        : bits(64),
+                     level     : nat)       -> bits(64) = {
+    let lsb  : nat       = PAGESIZE_BITS + level * sv_params.vpn_size_bits;
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.vpn_size_bits) - 1;
+    ((va >> lsb) & mask)
+}
+
+// PRIVATE: Extract offset within page from VA
+function offset_of_va(va : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << PAGESIZE_BITS) - 1;
+    va & mask
+}
+
+// Valid xlen-wide values containing virtual addrs must have upper
+// bits equal to the MSB of the virtual address.
+// Virtual address widths depend on the virtual memory mode.
+// PRIVATE
+function is_valid_vAddr(struct { va_size_bits, _ } : SV_Params,
+                        vAddr                      : bits(64))  -> bool = {
+  let index      = va_size_bits - 1;
+  let upper_bits = vAddr[63 .. va_size_bits];
+  let va_msb     = vAddr[index];
+  if va_msb == bitzero then
+    upper_bits == zeros()
+  else
+    upper_bits == ones()
+}
+
+// ****************************************************************
+// PTE (Page Table Entry) in PTN (Page Table Node)
+
+// PTE      MSBs      PPNs      RSW    BITs
+// Sv32      -       31..10    9..8    7..0
+// Sv39    63..54    53..10    9..8    7..0
+// Sv48    63..54    53..10    9..8    7..0
+
+// MSBs of PTE are reserved for RV64 extensions.
+// There are no available bits on RV32, so these bits will be zeros on RV32.
+
+// For PTW extensions (non-standard)
+type extPte = bits(64)
+
+// PRIVATE: extract msbs of PTE above the PPN
+function msbs_of_PTE(sv_params : SV_Params, pte : bits(64)) -> bits(64) = {
+  let mask : bits(64) = (zero_extend(0b1) << sv_params.pte_msbs_size_bits) - zero_extend(0b1);
+  (pte >> sv_params.pte_msbs_lsb_index) & mask
+}
+
+// PRIVATE: extract PPNs of PTE
+function PPNs_of_PTE(sv_params : SV_Params, pte : bits(64)) -> bits(64) = {
+  let mask : bits(64) = (zero_extend(0b1) << sv_params.pte_PPNs_size_bits) - zero_extend(0b1);
+  (pte >> sv_params.pte_PPNs_lsb_index) & mask
+}
+
+// PRIVATE: extract LSBs (including permissions) of PTE
+function D_of_PTE(pte : bits(64)) -> bit = pte[7]
```

Comment:
The original code had three separate bitfields for PTEs for Sv32, Sv39 and Sv48, much of the PTEs are different for each one.  To unify these (since bitfields are not allowed to be polymorphic at the moment), I resorted to working at 64bits throughout and using selector functions, and using SV_Params to parameterize the differences.

Of course, the bottom 8 bits of PTEs are the same for Sv32, Sv39 and Sv48, and one could use a common bitfield for that part of the PTE, but I’m not sure it would clarify things much.

Reply:
I think using a bitfield for the common bits would be slightly easier to read than these wrapper functions.

Reply:
Ok, I’ll defer to the majority here and do the common 8 LSBs with a bitfield, although IMHO while bitfields are a nice construct, I don’t think it  improves readability in this particular situation, especially since other fields in the PTE are not done this way.

Reply:
Github won't let me actually resolve this for some reason but it's resolved.